src/blob/index/simple.rs

```diff
@@ -1,377 +1,310 @@
-            if let Some(v) = headers.get_mut(header.key()) {
-                v.push(header)
-            } else {
-                headers.insert(header.key().to_vec(), vec![header]);
-            }
-            Ok(headers)
-        })
+    async fn get_records_headers(&self) -> Result<(InMemoryIndex, usize)> {
+        let mut buf = self.file.read_all().await?;
+        self.validate_header(&mut buf).await?;
+        let offset = self.header.meta_size + self.header.serialized_size()? as usize;
+        let records_buf = &buf[offset..];
+        (0..self.header.records_count)
+            .try_fold(InMemoryIndex::new(), |mut headers, i| {
+                let offset = i * self.header.record_header_size;
+                let header: RecordHeader = deserialize(&records_buf[offset..])?;
+                // We used get mut instead of entry(..).or_insert(..) because in second case we
+                // need to clone key everytime. Whereas in first case - only if we insert new
+                // entry.
+                if let Some(v) = headers.get_mut(header.key()) {
+                    v.push(header)
+                } else {
+                    headers.insert(header.key().to_vec(), vec![header]);
+                }
+                Ok(headers)
+            })
+            .map(|headers| (headers, self.header.records_count))
     }
 
-    async fn dump_in_memory(&mut self, buf: Vec<u8>) -> Result<usize> {
-        let _ = std::fs::remove_file(self.name.to_path());
-        let file = File::create(self.name.to_path(), self.ioring.clone())
+    async fn get_any(&self, key: &[u8]) -> Result<Option<RecordHeader>> {
+        Self::binary_search(&self.file, key, &self.header)
             .await
-            .with_context(|| format!("file open failed {:?}", self.name.to_path()))?;
-        let size = file.write_append(&buf).await?;
-        self.apply_written_byte(&file).await?;
-        file.fsyncdata().await?;
-        self.inner = State::OnDisk(file);
-        self.mem = None;
-        Ok(size)
-    }
-
-    async fn apply_written_byte(&mut self, file: &File) -> Result<()> {
-        self.header.written = 1;
-        let serialized = serialize(&self.header)?;
-        file.write_at(0, &serialized).await?;
-        Ok(())
+            .map(|res| res.map(|h| h.0))
     }
 
-    async fn load_in_memory(&mut self, file: File) -> Result<()> {
-        let mut buf = file.read_all().await?;
-        trace!("read total {} bytes", buf.len());
-        let header = Self::deserialize_header(&buf)?;
-        if header.written != 1 {
-            return Err(Error::validation("header was not written").into());
-        }
-        if !Self::hash_valid(&header, &mut buf)? {
-            return Err(Error::validation("header hash mismatch").into());
-        }
-        if header.version != HEADER_VERSION {
-            return Err(Error::validation("header version mismatch").into());
+    fn validate(&self) -> Result<()> {
+        if self.header.written == 1 {
```

Comment:
As I mentioned above, this function only checks written byte because I need the whole file content to check hash (but when we start Pearl from file, we don't load that content, we load only header in RAM, so I check written byte in header in this case)

Reply:
I left logic, which was in simple index implementation

Reply:
Added FIXME: as this is a part of old logic I'll leave it like this, but you'll have a notice, that maybe this should be changed.